compiler/sem/passes.nim

```diff
@@ -176,39 +208,99 @@ proc processModule*(graph: ModuleGraph; module: PSym; idgen: IdGenerator;
       # for the interactive mode.
       if module.name.s != "nimscriptapi":
         processImplicits(
-          graph, graph.config.active.implicitImports, nkImportStmt, a, module)
+          graph, graph.config.active.implicitImports,
+          nkImportStmt, passesArray, module)
+
         processImplicits(
-          graph, graph.config.active.implicitIncludes, nkIncludeStmt, a, module)
+          graph, graph.config.active.implicitIncludes,
+          nkIncludeStmt, passesArray, module)
+
+    # Until toplevel compilation fails (returns `false` from processing),
+    # execute the compilation
+    var processingOk = true
+    while processingOk:
+      # Processing was ok but compilation was halted via something else.
+      if graph.stopCompile():
+        break
 
-    while true:
-      if graph.stopCompile(): break
-      var n = parseTopLevelStmt(p)
-      if n.kind == nkEmpty: break
-      if n.kind in imperativeCode:
+      # Get next 'first' statement in the file
+      var firstStatement = parser.parseTopLevelStmt().toPNode()
+      if firstStatement.kind == nkEmpty:
+        break
+
+      # if this is some kind of imperative code - collect subsequent
+      # statements as well
+      #
+      # TODO:BUG despite looking similar feature-wise these two branches
+      # actually process the code in different manner - there is a BUG in
+      # phase ordering which makes the following scenario possible:
+      # sequence of imperative code blocks `I2 I1 I3` is concatenated in
+      # one group and processed by the semantic pass in one go. `when` is
+      # also considered an imperative node kind, which means conditional
+      # declarations are concatenated together. In most use cases it is not
+      # particularly problemantic, but when declaration of the symbols such
+      # as `nimGCvisit` is concerned this means if `I1` defines symbol and
+      # `I2` uses it *in the backend* grouping enables this code to be
+      # processed:
+      #
+      # - `I2` is compiled, does not need  in sem
+      # - `I1` defines symbol
+      # - `I2` in `[I2, I1, I3]` needs  in backend -  is present
+      #
+      # without ordering
+      #
+      # - `I2` is compiled, does not need symbol in sem
+      # - `I2` needs symbol in backend - it is not present yet => fail
+      #
+      # This can be reproduced by disabling first branch and compiling any
+      # file - should fail in `system.nim` processing with `system needs
+      # 'nimGCvisit'` error.
+      if firstStatement.kind in imperativeCode:
         # read everything until the next proc declaration etc.
-        var sl = newNodeI(nkStmtList, n.info)
-        sl.add n
+        var toplevelStatements = newNodeI(nkStmtList, firstStatement.info)
+        toplevelStatements.add firstStatement
+
+        # Tail data that is would have already been parsed by the time we
+        # collect all the necessary statements - it should be processed as
+        # a separate trailing action.
```

Comment:
```suggestion
```

I don't think this is accurate, `rest` stores everything that's not part of this collection but was parsed because we need to parse and then decide whether to continue.

I added a suggestion which documents the rest var itself, see what you think below.